### PR TITLE
fix: Add retry for installation

### PR DESCRIPTION
**Pull request by razbensimon**
**Issue:**
installation is flaky (mostly version 0.35.0)
```
curl: (22) The requested URL returned error: 503

tgenv: tgenv-install: [0;31m[ERROR] Tarball download failed[0;39m
```

**Solution:**
- **add a retry function for installation**
    - support calling with whatever arguments 
    - command could be a bash function OR script file/binary
    - optional arguments support (`sleep`, `tries`)
    - special argument `stop-on-exit-code` for making condition when NOT to retry. (use for 4XX)
- add retry function for `use` script
	- this is a bonus, I had flaky errors with that too (`Something is seriously wrong` error)
- Added appropriate logs

**More explanation:**
-  Before the change
	- each 4XX/5XX curl failure caused curl to fail with exit code 22 (by `-f` flag)
	- no retry
- After the change
    - retry on 5XX or exit code != 0
    - DO not retry on 4XX (user's fault)

**QA:**

[x] retry on 5XX
image

[x] Do not retry on 4XX (invalid version for example)
image

[x] retry on curl exit code != 0 
image

[x] Success installation + `use` script
image

### CODE REVIEW

**Comment by razbensimon** [comment]
[tgenv installation is flaky](https://trello.com/c/LfXIPIni/9911-tgenv-installation-is-flaky)

File: libexec/tgenv-install [hunk 1]
```diff
@@ -60,19 +61,38 @@ tarball_name="terragrunt_${os}";
...
+}
+
+
 mkdir -p "${dst_path}" || error_and_die "Failed to make directory ${dst_path}"
 
 info "Downloading release tarball from ${version_url}/${tarball_name}"
-curlw -# --location-trusted -f -o "${dst_path}/terragrunt" "${version_url}/${tarball_name}"
-curl_status=$?
-if [ "$curl_status" -eq 0 ]; then
+retry -tries 5 -sleep 5 -stop-on-exit-code 2 install_tg_version || error_and_die "Tarball download failed"
```
Comment (ronnyorot) [inline libexec/tgenv-install:86]: won't this `error_and_die` here will just make the code stop running, and we won't get to the cleanup `else` block (lines 96-97) at all?

Comment (razbensimon) [reply libexec/tgenv-install:86]: True, ill fix that wnd probbaly will move 
```
# make sure versions folder does not exist, so we can retry installation
  rm -rf "${dst_path}"
```

to `install_tg_version`

File: libexec/tgenv-install [hunk 2]
```diff
@@ -60,19 +61,44 @@ tarball_name="terragrunt_${os}";
 
 info "Installing Terragrunt v${version}"
 
+function install_tg_version() {
+  local http_code
+  http_code=$(curlw -# --location-trusted -w "%{http_code}" -o "${dst_path}/terragrunt" "${version_url}/${tarball_name}")
```
Comment (razbensimon) [inline libexec/tgenv-install:66]: FYI this helped me to QA mock:
```bash
  http_code=$(curlw -s --write-out "%{http_code}" --output /dev/null http://httpbin.org/status/500)
```

File: libexec/tgenv-install (hunk 1 above)
Comment (razbensimon) [reply libexec/tgenv-install:95]: @ronnyorot [fix folder cleanup after failed install](https://github.com/env0/tgenv/pull/11/commits/0cb0f396cb358c7f16fe023c722f718d7d7ee495)

great catch!

**Review by ghost — APPROVED** [review]
Great job!

**Pull request merged**